Add node relocation functions.

seigodev opened the pull request:
ノードの重心を計算し、重心が小さいものから順に並ぶようにする関数を追加。
sort_nodes_by_barycenter_top_to_bottom(): ノードを重心が小さい順に並び変える。上の階層から下の階層へと操作を行う。
part_nodes_par_level(): ノードを階層ごとに分ける。
calc_node_barycenter_at_targets(): target_nodesから各ノードの重心を計算する。

seigodev committed:
Add node relocation functions.

ノードの重心を計算し、重心が小さいものから順に並ぶようにする関数を追加。
sort_nodes_by_barycenter_top_to_bottom(): ノードを重心が小さい順に並び変える。上の階層から下の階層へと操作を行う。
part_nodes_par_level(): ノードを階層ごとに分ける。
calc_node_barycenter_at_targets(): target_nodesから各ノードの重心を計算する。

aabaa commented on demo/create_graph.py:
```diff
@@ -266,6 +266,81 @@ def cut_edge(source, target):
+            barycenter_dict[node] = 1000
+        else:
+            barycenter_dict[node] = barycenter_sum / len(node.target_nodes)
+    return barycenter_dict
```
関数名がしっくりくるように関数を分けます
```python
def calc_xcenter_from_targets(node):
    if len(node.targets) > 0:
        return sum(node.targets) / len(node.targets)
    else:
        return 1000

def create_node2xcenter_dict(nodes):
    return {node: calc_xcenter_from_targets(node) for node in nodes}
```

aabaa commented on demo/create_graph.py:
```diff
@@ -266,6 +266,81 @@ def cut_edge(source, target):
+        if node.y not in each_level_nodes:
+            each_level_nodes[node.y] = []
+        each_level_nodes[node.y].append(node)
+    return each_level_nodes
```
関数名を考えるときに英和辞典，類語辞典，Google翻訳等を使ってください
```python
def divide_nodes_by_level(nodes):
    return {node.y: node for node in nodes}
```

aabaa commented on demo/create_graph.py:
```diff
@@ -266,6 +266,81 @@ def cut_edge(source, target):
+        sort_level_nodes = [node for node, barycenter in barycenter_sorted]
+        # 順にxを割り振る。
+        assign_x_coordinate(sort_level_nodes)
+        sort_level += 1
```
```python
def sort_nodes_by_xcenter_from_top_to_bottom(all_nodes):
    level2nodes = divide_nodes_by_level(all_nodes)
    for level, nodes in sorted(level2nodes.items()):
        node2xcenter = create_node2xcenter_dict(nodes)
        sorted_items = sorted(node2xcenter.items(), key=lambda x: x[1])
        sorted_nodes = list(sorted_items.keys())
        assign_x_coordinate(sorted_nodes)
```

seigodev committed:
Update create_graph.py

269~344行：交差削減のコードの追加
・sort_nodes_by_xcenter()：各ノードを重心でソート
・divide_nodes_by_level()：階層をキーに、値をその階層のノードとする辞書の作成
・node2xcenter()：各ノードに重心を与える
・calc_xcenter()：ノードの重心を計算する
・assign_x_by_xcenter()：x座標を重心をもとに割り当て。
175, 443, 445行：assign_x_coordinate() -> assign_x_sequentially()
その他：Nodeオブジェクトの属性名の変更
・target_nodes -> targets
・source_nodes -> sources

aabaa commented on demo/create_graph.py:
```diff
@@ -242,30 +242,113 @@ def cut_edge(source, target):
     return dummy
 
 
+def sort_nodes_by_xcenter(all_nodes, downwards):
```
downwards -> downward (sは不要)

aabaa commented on demo/create_graph.py:
```diff
@@ -242,30 +242,113 @@ def cut_edge(source, target):
+        v1=Nodeオブジェクト、v2=v1の重心の値(float)
+    Args:
+        nodes:重心を求めたいNodeオブジェクトのリスト。Nodeオブジェクトの階層は等しいのが好ましい。
+        from_targets: True:重心をtargetsを用いて計算する, False:重心をsourceesを用いて計算する。
```
sourcees -> sources

aabaa commented on demo/create_graph.py:
```diff
@@ -242,30 +242,113 @@ def cut_edge(source, target):
+            v1: Nodeオブジェクト
+            v2: 重心の値(float)
+    """
+    return [(node, calc_xcenter(node.targets if from_targets else node.sources)) for node in nodes]
```
何でも一行にする必要はないです．
三項演算子の箇所は行を分けると良いでしょう．
```python
calc_nodes = node.targets if from_targets else node.sources
return [(node, calc_xcenter(calc_nodes) for node in nodes]
```

aabaa commented on demo/create_graph.py:
```diff
@@ -242,30 +242,113 @@ def cut_edge(source, target):
+    Return:
+        重心の値(float)
+    """
+    return sum([node.x for node in nodes]) / len(nodes) if len(nodes) > 0 else float('infinity')
```
フラグのON/OFF設定より多くの処理を三項演算子で行うことは可読性が下がるので避けましょう．
```python
if len(nodes) > 0:
    return sum([node.x for node in nodes])
else:
    return float('infinity')
```

seigodev committed:
Fix requested changes.

・downwards -> downward
・sourcees -> sources
・node2xcenter()の変更(314-317)
・calc_xcenter()の変更(333-336)
